src/lib.rs

```diff
@@ -150,110 +213,260 @@ impl<'object, Endian> DebugInfo<'object, Endian>
             let rows = unit.line_rows(&self.debug_line);
             if rows.is_err() {
-                continue;
+                return None;
             }
-
             let mut rows = rows.unwrap();
-            while let Ok(Some((header, row))) = rows.next_row() {
-                if row.address() > addr {
-                    if let Some(ref row) = current {
-                        let mut path = String::new();
-                        let file = row.file(header).unwrap();
-                        if let Some(directory) = file.directory(header) {
-                            let directory = directory.to_string_lossy();
-                            if !directory.starts_with("/") {
-                                if let Some(comp_dir) = unit.comp_dir() {
-                                    path.push_str(&*comp_dir.to_string_lossy());
-                                    path.push_str("/");
-                                }
-                            }
-                            path.push_str(&*directory);
-                            path.push_str("/");
-                        }
-                        path.push_str(&*file.path_name().to_string_lossy());
 
-                        return Some((path, row.line().unwrap()));
+            // Now, find the last row before a row with a higher address than the one we seek.
+            let mut current = None;
+            while let Ok(Some((_, row))) = rows.next_row() {
+                if row.address() <= addr {
+                    if row.end_sequence() {
+                        current = None;
+                    } else {
+                        // Might be the right row, but we won't know until we see the next one.
+                        // The .clone is needed so we can keep iterating
+                        current = Some(row.clone());
+                    }
+                    continue;
+                }
+                break;
+            }
+
+            // The row we just last iterated to is *after* the address, to the previous row we
+            // saw (stored in current) is the one we want. There better be one...
+            let row = current.unwrap();
+            let header = rows.header();
+
+            let mut path = path::PathBuf::new();
+            let file = row.file(&header).unwrap();
+            if let Some(directory) = file.directory(&header) {
+                let directory = directory.to_string_lossy();
+                if !directory.starts_with("/") {
+                    if let Some(comp_dir) = unit.comp_dir() {
+                        path.push(&*comp_dir.to_string_lossy());
                     }
-                    break;
                 }
-                if row.end_sequence() {
-                    current = None;
+                path.push(&*directory);
+            }
+            path.push(&*file.path_name().to_string_lossy());
+
+            let line = row.line().unwrap();
```

Comment:
Why `unwrap` here? We should be able to handle an unknown line number and print `??`, right?

I guess we did an `unwrap` in the old code too. It's pretty clear that it was an example program, and not as polished as `gimli` proper :P